python_modules/dagster/dagster/_config/structured_config.py

```diff
@@ -15,6 +21,88 @@ class Config(BaseModel):
     """
 
 
+class cached_method:
+    """
+    Caches an instance method on a structured config class using :py:method:`functools.lru_cache`.
+    """
+
+    def __init__(self, function):
+        self.function = lru_cache()(function)
+
+    def __call__(self, *args, **kwargs):
+        return self.function(*args, **kwargs)
+
+    def __get__(self, instance, _):
+        from functools import partial
+
+        return partial(self.__call__, instance)
+
+
+class Resource(
+    ResourceDefinition,
+    Config,
+):
+    """
+    Base class for Dagster resources that utilize structured config.
+
+    This class is a subclass of both :py:class:`ResourceDefinition` and :py:class:`Config`, and
+    provides a default implementation of the resource_fn that returns the resource itself.
+
+    Example:
+
+    .. code-block:: python
+
+        class WriterResource(Resource):
+            prefix: str
+
+            def output(self, text: str) -> None:
+                print(f"{self.prefix}{text}")
+
+    """
+
+    class Config:
+        arbitrary_types_allowed = True
+        keep_untouched = (cached_property, cached_method)
+        frozen = True
```

Comment:
is this supposed to be pydantic.Config?

If not this is super confusing and worth of comment.

Reply:
I see looks like Pydantic looks for a bare class called Config by convention and there is no way for us to avoid the name collison.

At a minimum there should be a comment describing this.